app/src/main/java/de/westnordost/streetcomplete/data/quest/VisibleQuestsSource.kt

```diff
@@ -120,12 +131,21 @@ class VisibleQuestsSource(
         listeners.remove(listener)
     }
 
+    fun clearCache() = cache.clear()
+
+    fun trimCache() = cache.trim(SPATIAL_CACHE_SIZE / 3)
+
     private fun updateVisibleQuests(addedQuests: Collection<Quest>, deletedQuestKeys: Collection<QuestKey>) {
         if (addedQuests.isEmpty() && deletedQuestKeys.isEmpty()) return
+        cache.removeAll(deletedQuestKeys)
+        addedQuests.forEach(cache::putIfTileExists)
```

Comment:
This must be synchronized. I suggest to change the interface of `SpatialCache` to have a method like `update(updatedOrAdded: Iterable<T>, deleted: Iterable<K>)`

Might be considerably faster too because now for every single added quest, one synchronization block is entered and exited.